File: dump.py
import frida
import pefile
import lief
import pathlib
import binascii

class DUMPER:
    def __init__(self,process_name):
        self.process = process_name
        self.frida_script = "dump.js"
        self.frida_api = self.init_frida()
        self.pe_data = self.get_pe_from_mem()
        self.binary = self.init_lief()

    def init_frida(self):
        session = frida.attach(self.process)
        script_data = open(self.frida_script,"r").read()
        script = session.create_script(script_data)
        script.load()
        api = script.exports_sync
        api.init(self.process)
        return api
    
    def get_pe_from_mem(self):
        filesize = self.frida_api.get_file_size()
        pe_buf = bytearray(filesize)
        data = bytearray(self.frida_api.get_file_header())
        pe_buf[:len(data)]=data
        pe_hdr = pefile.PE(data=data)
        for section in pe_hdr.sections:
            section = section.dump_dict()
            rva = section['VirtualAddress']['Value']
            virtsize = section['Misc_VirtualSize']['Value']
            section_data = self.frida_api.get_mem(rva,virtsize)
            poffset = section['PointerToRawData']['Value']
            psize = section['SizeOfRawData']['Value']
            pe_buf[poffset:poffset+len(section_data)]=bytearray(section_data)
         
        return bytes(pe_buf)
    
    def init_lief(self):
        return lief.parse(self.pe_data)
    
    def write(self,path):
        builder = lief.PE.Builder(self.binary)
        builder.build_imports(True)
        builder.build()
        builder.write(path)

if __name__ == "__main__":
    dumper = DUMPER(r"notepad++.exe")
    dumper.write(r"C:\out.exe")

File: dump.js
rpc.exports = {
    init: function (target) {
        dump = new DUMPER(target);
    },
    getMem: function (rva,size) {
        if(dump){
            return dump.get_mem(rva,size);
        }
        return 0
    },
    getFileSize(){
        if(dump){
            return dump.size;
        }
        return 0;
    },
    getFileHeader(){
        if(dump){
            return dump.header;
        }
        return 0;
    }
};

//init with a target process name
var dump;

class DUMPER {
    target;
    constructor(target){
        this.target = target;
    }

    get header(){
        const pe_offset = this.base.add(0x3c).readInt();
        const size_headers = this.base.add(pe_offset).add(0x54).readInt()
        const data = this.base.readByteArray(size_headers);
        return data
    }

    get size(){
        return Process.getModuleByName(this.target).size;
    }

    get base(){
        return Process.getModuleByName(this.target).base;
    }

    rva2va(rva){
        return this.base.add(ptr(rva));
    }

    get_mem(rva,size){
        const mem_ptr = this.rva2va(rva);
        const range = Process.getRangeByAddress(mem_ptr);
        if(!range.protection.includes("r")){
            Memory.protection(range.base,range.size,"rwx");
        }
        return mem_ptr.readByteArray(size);
    }
}

